src/merge.rs

```diff
@@ -298,30 +319,56 @@ impl <'t, D: Driver> Merger<'t, D> {
     fn two_way_merge(&mut self,
                      local_node: Node<'t>,
                      remote_node: Node<'t>)
-                     -> Result<MergedNode<'t>>
+                     -> Result<TwoWayMergeResult<'t>>
     {
         trace!(self.driver, "Item exists locally as {} and remotely as {}",
                local_node,
                remote_node);
 
-        if !local_node.has_compatible_kind(&remote_node) {
-            // TODO(lina): Remove and replace items with mismatched kinds in
-            // `check_for_{}_structure_change_of_{}_node`.
-            error!(self.driver, "Merging local {} and remote {} with different kinds",
-                   local_node, remote_node);
-            return Err(ErrorKind::MismatchedItemKind(local_node.kind, remote_node.kind).into());
-        }
-
+        self.merged_guids.insert(local_node.guid.clone());
         self.merged_guids.insert(remote_node.guid.clone());
 
-        if local_node.guid != remote_node.guid {
-            // We deduped a NEW local item to a remote item.
-            self.merged_guids.insert(local_node.guid.clone());
+        if local_node.has_compatible_kind(&remote_node) {
+            self.merge_one_node(local_node, remote_node)
+        } else {
+            // For mismatched kinds, we keep both nodes with new GUIDs. This is
+            // an edge case: Places maintenance on Desktop can change an item's
+            // kind without changing its GUID (bug XXXXXXX), but this should be
```

Comment:
[Here's](https://searchfox.org/mozilla-central/rev/e6520f0a4dd5d7474c32a1164744953ea59be0d0/toolkit/components/places/PlacesDBUtils.jsm#465,491) where we change bookmarks without URLs to folders, and anything with a URLs to bookmarks. I should file a bug so that we also update the GUID when we do that.

Reply:
Yeah, probably. As a note, I'm somewhat unconvinced that keeping both items is really the right call here. Yes, it avoids data loss, but these are items that are fairly corrupted to begin with.

That said, I'm not sure how that would actually work out, so we should probably just defer to your judgement here :)